ICU-21310 Minimize ICU4C and ICU4J Readmes

echeran opened the pull request:
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21310
- [X] Required: The PR title must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [ ] Required: Each commit message must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [X] API docs and/or User Guide docs changed or added, if applicable

The PR does a few things:
* Replace the contents of the readme.html files for ICU4C and ICU4J with a pointer to the user guide
* Re-organize user guide content that is ICU4C- and ICU4J-specific to be together
* Pull out the ICU4C build instructions to a separate page. Likewise, consolidate the packaging instructions with the existing packaging page.

My personal fork is configured to render the branch's changes, at https://echeran.github.io/icu/

markusicu commented:
Sorry I am late -- will review today.

jefgen reviewed (commented):
Ah, sorry for not getting a chance to take a look at this sooner...
FWIW, I think the changes look good; I just had a few minor comments/suggestions.
Thanks for doing all this work @echeran! :)

markusicu commented on docs/userguide/icu/icu4c-readme/icu4c-build.md:
```diff
@@ -1,15 +1,16 @@
-parent: ICU
+title: Building ICU4C
+nav_order: 2
+parent: ICU4C Readme
```
I think this reorg makes sense. However, I wonder if we could reduce the depth.

How about, both in the navbar and in file folders,
moving icu/icu4c-readme/... "ICU > ICU4C Readme" one level up to icu4c/... "ICU4C > Readme"
and same with icu/icu4j-readme/... "ICU > ICU4J Readme" one level up to icu4j/... "ICU4J > Readme"
?

And in these icu4c/... and icu4j/... folders I would reduce path-filename redundancy, e.g., icu4c/icu4c-build.md --> icu4c/build.md etc.

What do you think?

markusicu commented on docs/userguide/icu/icu4j-readme/index.md:
```diff
@@ -1,8 +1,9 @@
+nav_order: 8
 parent: ICU
+has_children: true
 ---
```
```
docs/userguide/icudata.md → docs/userguide/icu/icudata.md
```
I can't figure out how to comment on a file move without contents changes...

Shouldn't this move to icu_data/index.md?

echeran replied:
Done. Yes, thanks.

markusicu commented on docs/userguide/icu/icu4c-readme/index.md:
```diff
@@ -0,0 +1,448 @@
+| Online ICU Demos                                   | <http://demo.icu-project.org/icu-bin/icudemos>                           |
+| Contacts and Bug Reports/Feature Requests          | <http://site.icu-project.org/contacts>                                   |
+
+**Important:** Please make sure you understand the [Copyright and License Information](http://source.icu-project.org/repos/icu/trunk/icu4c/LICENSE).
```
Yes, this is a broken svn link.

markusicu commented on docs/userguide/icu/icu4c-readme/index.md:
```diff
@@ -47,14 +47,15 @@ The ICU homepage also links to related information about writing internationaliz
-| ICU4C API Documentation Online             | <http://icu-project.org/apiref/icu4c/>                         |
-| Online ICU Demos                           | <http://demo.icu-project.org/icu-bin/icudemos>                 |
-| Contacts and Bug Reports/Feature Requests  | <http://site.icu-project.org/contacts>                         |
+| ICU, ICU4C & ICU4J Homepage                        | <http://icu-project.org/>                                                |
```
http://site.icu-project.org/

markusicu reviewed (commented):
Hi @echeran , first of all, this is a nice piece of work, thank you!

I have a suggestion for a slightly different (less deep) reorg of some of the moved pages.

echeran commented on docs/userguide/icu/icu4c-readme/index.md:
```diff
@@ -0,0 +1,448 @@
+
+Since `u_init()` will load only one or two data files, it cannot guarantee that all of the data that an application needs is available. It cannot check for all data files because the set of files is customizable, and some ICU services work without loading any data at all. An application should always check for error codes when opening ICU service objects (using `ucnv_open()`, `ucol_open()`, C++ c ...
+
+#### ICU 3.4 and later
```
Right. This markdown content originally came from the ICU4C readme.html. That conversion was done in the previous PR a few months ago.

What is changing here of interest in this PR is that I "refactored" out the deeply nested sections on building ICU4C into a separate sub-page. But this PR is also moving files and changing the directory structure, and the combination seems to be enough to throw off the git diff, which is why it looks like this entire page is added.

echeran commented:
Thanks @jefgen @markusicu for good feedback. I applied it all in the commit [62627e7c](https://github.com/unicode-org/icu/pull/1636/commits/6267e7c8e1798bd4b4d1948488402cd693d162c1). To wit:
* Applied changes to flatten structure so that ICU4C and ICU4J sections, readmes and subpages are siblings of the ICU section
  - Updated links in the content and config values in markdown headers accordingly
* Moved `icu/icudata.md` to `icu_data/index.md`
* Applied fixes to broken links and wording

FYI -- in order to better preserve the comments within the Github PR diff view, I tried a different way to resolve the merge conflict instead of the usual rebase on latest changes. Instead, I merged the latest changes into the PR branch, and addressed the merge conflicts within the merge commit that gets added at the end on the feature branch.

echeran commented:
I pushed two more commits. One is to fix merge conflicts after we renamed our main branch to `main`. The other is catching changes not made in previous commits for the ICU4C and ICU4J sections flattening.
Demo site updated: https://echeran.github.io/icu/

markusicu commented on docs/userguide/icu/icu4c-readme/index.md:
```diff
@@ -0,0 +1,448 @@
+
+The subpage for the current release will also include an API Change Report, both for ICU4C and ICU4J, for a complete list of APIs added, removed, or changed in this release.
+
+The list of API changes since the previous ICU4C release is available [here](https://htmlpreview.github.io/?https://raw.githubusercontent.com/unicode-org/icu/master/icu4c/APIChangeReport.html).
```
Does not look done, but I think it can wait for a follow-up PR. Maybe Jeff can propose wording then ;-)

markusicu commented on docs/userguide/icu/icu4c-readme/index.md:
```diff
@@ -0,0 +1,448 @@
+
+### Using ICU in a Multithreaded Environment
+
+Some versions of ICU require calling the `u_init()` function from `uclean.h` to ensure that ICU is initialized properly. In those ICU versions, `u_init()` must be called before ICU is used from multiple threads. There is no harm in calling `u_init()` in a single-threaded application, on a single-CPU machine, or in other cases where `u_init()` is not required.
```
Good suggestion, but it looks like this is not done yet. Jeff, follow-up PR?

markusicu commented:
> FYI -- in order to better preserve the comments within the Github PR diff view, I tried a different way to resolve the merge conflict instead of the usual rebase on latest changes. Instead, I merged the latest changes into the PR branch, and addressed the merge conflicts within the merge commit that gets added at the end on the feature branch.

Yes, it seems to help, although GitHub is still not good with showing comments in context.

FYI Shane actually documented this technique: http://site.icu-project.org/repository/gitdev under "Now, resolve the conflicts."

markusicu commented on docs/userguide/icu/editing.md:
````diff
@@ -159,6 +159,7 @@ To activate the version of Ruby:
 ```bash
 rbenv init  # OR:  eval "$(rbenv init -)"
 rbenv shell <version-num>
+rbenv versions  # verify the specified version is in use
````
Whole-file comment: From the navbar grouping (which makes sense to me), it looks like this should not be at icu/editing.md but rather at dev/editing.md

markusicu replied:
Comment about a different file:
I see you moving the glossary into the icu/ folder, but in the navbar it's still at top level. They should be consistent; not sure which is better. I am ok with moving it in the navbar under ICU.

markusicu commented:
FYI After this PR I want to take one more look at what I wrote in https://unicode-org.atlassian.net/browse/ICU-21310 and see if it makes sense to move more contents and/or connect the various readmes differently etc.

markusicu reviewed (dismissed):
lgtm tnx

jefgen reviewed (approved):
lgtm. Thank you very much for doing this @echeran!

markusicu reviewed (approved):
lgtm
